File: solutions/go/src/playground/algorithm/spiral_matrix.go
package algorithm

func spiralOrder(matrix [][]int) []int {
	if len(matrix) == 0 {
		return nil
	}

	result := make([]int, 0)
	up, down, left, right := 0, len(matrix)-1, 0, len(matrix[0])-1

	for left <= right || up <= down {
		for l := left; up <= down && l <= right; l++ {
			result = append(result, matrix[up][l])
		}
		up++

		for u := up; left <= right && u <= down; u++ {
			result = append(result, matrix[u][right])
		}
		right--

		for r := right; up <= down && r >= left; r-- {
			result = append(result, matrix[down][r])
		}
		down--

		for d := down; left <= right && d >= up; d-- {
			result = append(result, matrix[d][left])
		}
		left++
	}

	return result
}

File: solutions/java/src/main/java/playground/algorithm/SpiralMatrix.java
package playground.algorithm;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class SpiralMatrix {

    public List<Integer> spiralOrder(int[][] matrix) {
        if (matrix.length < 1) {
            return Collections.emptyList();
        }
        return spiralOrder(matrix, 0, matrix[0].length - 1, 0, matrix.length - 1);
    }

    private List<Integer> spiralOrder(int[][] matrix, int startX, int endX, int startY, int endY) {
        List<Integer> result = new ArrayList<>();
        for (int x = startX; x <= endX && startY <= endY; x++) {
            result.add(matrix[startY][x]);
        }
        startY++;
        for (int y = startY; y <= endY && startX <= endX; y++) {
            result.add(matrix[y][endX]);
        }
        endX--;
        for (int x = endX; x >= startX && startY <= endY; x--) {
            result.add(matrix[endY][x]);
        }
        endY--;
        for (int y = endY; y >= startY && startX <= endX; y--) {
            result.add(matrix[y][startX]);
        }
        startX++;
        if (!result.isEmpty()) {
            result.addAll(spiralOrder(matrix, startX, endX, startY, endY));
        }
        return result;
    }
}
